### PR TITLE
Add more device and precision apis.

### PR DESCRIPTION
**Pull request by pschuh**
Update Optimizer, comparison ops BatchNorm and DifferentialOperators to handle mixed precision and devices.
The BatchNorm change is needed to reduce the number of zeros in the vjps.

### CODE REVIEW

File: Sources/TensorFlow/Core/Runtime.swift [hunk 1]
```diff
@@ -567,6 +567,15 @@ struct DeviceScopes {
     }
 }
 
+// Evaluate the pullback on a one.
+@usableFromInline
+func pullbackOfOneLikeY<T: TensorFlowFloatingPoint, R>(
```
Comment (dan-zheng) [inline Sources/TensorFlow/Core/Runtime.swift]: `pullbackOfOneLikeY` could use a better name. We can defer this.

File: Sources/TensorFlow/Core/Device.swift [hunk 2]
```diff
@@ -14,3 +14,10 @@ extension Tensor {
         }
     }
 }
+
+extension _Raw {
+  static func toDevice<T: TensorFlowScalar>(_ x: Tensor<T>, _ device: Device) -> Tensor<T> {
```
[the remark this answers is not in the document]
Comment (pschuh) [reply Sources/TensorFlow/Core/Device.swift]: Acknowledged

File: Sources/TensorFlow/Core/Runtime.swift (hunk 1 above)
Comment (pschuh) [reply Sources/TensorFlow/Core/Runtime.swift]: Acknowledged.
